### DISCUSSION TITLE
DIY 2400 DUPLETX no AUTO_WIFI_ON_INTERVAL Functionality w/ESP32-WROOM-32

### DISCUSSION THREAD

**Discussion by 2xWhiskey** [ExpressLRS/ExpressLRS#1856] [Q&A] [answered]
## Current Behavior

     Using ExpressLRS Configurator v1.5.0, flashing DIY 2400 DUPLETX onto an ESP32-WROOM-32 & SX1280 E28 hardware combo does not support the AUTO_WIFI_ON_INTERVAL where flashing DIY 2400 TX ESP32 SX1280 E28 does.

## Steps to Reproduce

1.  Build module using a ESP32-WROOM-32 and a E28-2G4M27S SX1280

2.  Flash either ExpressLRS 2.5.1 or 3.0

3.  Power on module and allow to sit for set AUTO_WIFI_ON_INTERVAL time.

4.  Device does not auto create it's own wifi network as expected.

## Possible Solution (Not obligatory)

     Perhaps DIY DUPLETX does not support this hardware.  If so initiate documentation for DUPLETX to inform the layman of compatible/supported hardware.  DUPLETX documentation appears to be unpublished.

## Details

     I've also been unsuccessful at getting this hardware/firmware combination to successfully communicate with my Taranis X-Lite Pro via UART0 on the internal RF module connection, further suggesting this combination's incompatibility.  To verify the module's integrity, I successfully tested it by connecting it via the external module bay and using the DIY 2400 TX ESP32 SX1280 E28 firmware.

     I'm trying to replace my Taranis X-Lite Pro's internal RF module and retain the external module bay's functionality for other modules.  The radio is hard configured to disable the S.PORT line when the internal module is powered on, preventing use of the DIY 2400 TX ESP32 SX1280 E28 firmware.

     Don't think so, no clue what that is.

     Help me flick one more proverbial booger at FrSky for wantonly abandoning protocols and hardware by garnering more attention of people using ELRS on their hardware.  And the ergonomics of the X-Lite are just amazing for smaller handed thumbers such as my self.

## Your Environment

    All at my desk over a period of a few days troubleshooting, researching and pestering smarter people on Discord.

- TX hardware:  DIY 2400 TX ESP32 SX1280 E28, using a custom PCB to connect to the handset's internal RF bay connector
- RX hardware:  N/A, haven't gotten there yet
- Handset model:  FrSky Taranis X-Lite Pro
- OpenTX version (including nightly number):  EdgeTX 2.7.1
- ExpressLRS version (TX & RX MUST MATCH):  3.0.0 & 2.5.1
- Packet Rate:  115k & 400k
- Telemetry Ratio: unable
- user_defines:  AUTO_WIFI_ON_INTERVAL

**Comment by iottrends** [comment]
have you looked into the pin mappings of your custom pcb.
since E28SX1280 is working the pin mapping. DIY_2400_TX_ESP32_E28.h .
However the pin mappings for **dupletx** is different:
#define GPIO_PIN_RCSIGNAL_RX    13
#define GPIO_PIN_RCSIGNAL_TX    13
#define GPIO_PIN_LED_WS2812     15
#define GPIO_PIN_FAN_EN         17

Compare above pins with the DIY_2400_TX_ESP32_E28.h
My guess is on flashing Dupletx your code is exiting before entering into bind WIFI, probably a pin and RF init issue.

**Reply by 2xWhiskey** [reply]
Where do you find the .h files?  I keep looking and looking, but all I ever find are .JSON files...

**Reply by pkendall64** [reply]
The .h files are for the old 2.x targets and the STM targets. If you want to look at them you would have to checkout a V2 tag and look in /src/include/target/...

**Reply by 2xWhiskey** [reply]
Ok I think it's all making more sense now.  I've been re and reflashing the same board after having soldered on the E28, and because I haven't properly assigned the ESP32's GPIOs prior to, the E28 is basically causing the ESP32 to not behave as expected.  So I think I need to detach the E28 (or start a new board) and assign GPIOs via the .h, flash, then solder on the E28.  And of coarse archive my special .h.

**Comment by 2xWhiskey** [comment]
Thank you for your advice.  My code experience is next to nil so it will take me a while to research and decipher it for my understanding.

**Comment by pkendall64** [comment, marked as answer]
If you flash 3.0 using vscode/platformio and choose the "Unified_EXP32_2400_TX_via_UART" targets to flash, it will ask after building which device it is that you configuring, just press enter.
It will boot straight into wifi mode and when you go to the page you can configure the hardware layout.
If you look in the repository at the file in the directory [src/hardware/TX/](https://github.com/ExpressLRS/ExpressLRS/blob/master/src/hardware/TX/) you will see the files that are used when choosing a target for defining the layout.
Once you have your device configured, you can download the json file from the device by going to url /hardware.json

**Reply by 2xWhiskey** [reply]
So I need to check and verify how the pins are assigned on my ESP32, do the number assignments in the target .JSON correlate to the board's pins? If so I can correct them in the .JSON? And does the pin count start at 0 or 1 in the .JSON?

PS this journey has me gotten me started into learning Arduino and vscode etc. So thank you for your patience and help.

**Reply by pkendall64** [reply]
The numbers in the JSON are the GPIO numbers which are normally the numbers printed on the development boards.

**Reply by 2xWhiskey** [reply]
Great, so [iottrends](https://github.com/iottrends) above suggested that the way my E28 is currently wired could possible causing the ESP32 to fail.  For ELRS 3.0, I think I need to detach the E28 before following your procedure.  And I guess I will need to archive my special .JSON for future updates to avoid E28 detachment and GPIO redefining every time I do a new flash.

**Reply by iottrends** [reply]
Dont Detach the board, just modify the pin mapping in the header files and compile ExLRA. So basically.
Step 1. Identify bellow pins on E28 Module.
**Lora reset
NSS
SCL
MISO
MOSI
TX_EN
RX_EN.**
Now look on ESP32, Where above said E28 pins are Connected, 
For example NSS pin on E28 should be connected to IO5 pin on ESP32.
// GPIO pin definitions
**#define GPIO_PIN_NSS            5 
#define GPIO_PIN_BUSY           21
#define GPIO_PIN_DIO1           4
#define GPIO_PIN_DIO2           22
#define GPIO_PIN_MOSI           23
#define GPIO_PIN_MISO           19
#define GPIO_PIN_SCK            18
#define GPIO_PIN_RST            14
#define GPIO_PIN_RX_ENABLE      27
#define GPIO_PIN_TX_ENABLE      26**
//Moving on, assuming you have got you pcb printed so there will be the schematic file also.
//up to this there shouldn't be a problem as all modules use the same pin mappings.

After this [bellow] you need to check.
#define GPIO_PIN_RCSIGNAL_RX    3
#define GPIO_PIN_RCSIGNAL_TX    1

#define GPIO_PIN_PA_PDET        35

// Backpack pins
#define GPIO_PIN_DEBUG_RX       16
#define GPIO_PIN_DEBUG_TX       17
#define GPIO_PIN_BACKPACK_EN    25
#define GPIO_PIN_BACKPACK_BOOT  15

**Reply by 2xWhiskey** [reply]
So I am currently stuck with ELRS **v3.0.0** as I don't know how to clone a copy of the older **2.5.1** to work with.  I am seeing what you are saying with the **DIY 2400 DupleTX.json**.  I've been able to identify that my schematic was connecting **_DIO2 via IO2_**, where the .json had assigned **_"radio_dio2": 33,_** on line 6.  Rather than modify the code, I think it would be best to reroute my schematic connection.

With this new understanding I can answer a question that I was holding back, how would I add a backpack function later on.  I can see how now!

I'll be reflashing soon, after I cut a trace and place a new jumper.  My fingers are crossed!

**Reply by 2xWhiskey** [reply]
I'm still getting nothing, no wifi network, no communication to the handset.  I've tried flashing both 2.5.1 and 3.0.0 via the Configurator.  Attached is a copy of my current wiring schematic.  I'm using the default .json and .h
![DIY X-Lite Internal ELRS](https://user-images.githubusercontent.com/113867350/192160944-3ee84480-3104-453b-af67-6f90947b3e70.jpg)

**Reply by pkendall64** [reply]
IO2 is a bad pin to be using as an input because it use used to control the voltage applied to the internal flash, and if it's low at startup then the ROM will apply the wrong voltage and the MCU will not boot the application!
Also, you have not said if you flashed with no device selected like I mentioned. If you do that then it should go directly to wifi mode and allow you configure the hardware layout.
If you did, then perhaps monitor the serial port during boot and see what it says, perhaps it's rebooting and there might be a clue there.

**Reply by 2xWhiskey** [reply]
Ok, interesting to note that most if not all the documented DIY modules that show a schematic are using IO2 for DIO2.

I misinterpreted your suggestion.  I've been trying to flash the whole time with a target selected.  Also, I've never used Visual Studio until this thread was made.  I've been following the procedure on the wiki, https://www.expresslrs.org/2.0/software/toolchain-install/, but I am lost at step 3.  I do not see a "Unified_ESP32_2400_TX_via_UART" to select, and if choose one of the existing targets, it just shows the file for editing.  When I got to the PlatformIO task bar to build, the target file window disappears.

I'm in Discord, X2 Whiskey, if you'd like to see what I am doing.

**Reply by pkendall64** [reply]
![Screen Shot 2022-09-26 at 10 35 35 AM](https://user-images.githubusercontent.com/512740/192166879-88f24ecb-4749-4ef3-a5aa-f3eec16f1533.png)
Try these steps in vscode/platformio

**Reply by 2xWhiskey** [reply]
I'm connected to the device via WiFi network.  Your patience is phenomenal!  I'll get to making the pin assignments now.

**Reply by pkendall64** [reply]
Awesome good job! You should be able to get some of the values for power output etc from the current E28 json file.

**Reply by 2xWhiskey** [reply]
Ok, still working on it.  Did you notice that the descriptor for DIO2 says it is unused on the SX128x?  Mine has been physically connected, perhaps should break it?  Good job on you, lol.

**Reply by pkendall64** [reply]
Yeah, I'd disconnect it just to be sure, but I know it's not used on SX1280. I'm the one who did the whole Unified target code and the hardware page 😄

**Reply by 2xWhiskey** [reply]
I think I saw a note that the purpose was to reduce target confused and the plethora of noob questions like mine.  Looking at my schematic, I don't have a connection for PDET pin or the PA enable pin, I assume I can leave those blank?

**Reply by pkendall64** [reply]
Yeah you can leave those blank. From your schematic above, it looks like this is all you need.
```
{
    "serial_rx": 3,
    "serial_tx": 1,
    "radio_busy": 21,
    "radio_dio1": 4,
    "radio_miso": 19,
    "radio_mosi": 23,
    "radio_nss": 5,
    "radio_rst": 14,
    "radio_sck": 18,
    "power_rxen": 27,
    "power_txen": 26,
    "power_min": 0,
    "power_high": 4,
    "power_max": 4,
    "power_default": 2,
    "power_control": 0,
    "power_values": [-15,-11,-8,-5,-1]
}
```
If you are adding a backpack later on, just ping me and I'll give you some information about how to wire it so it can be flashed by passing through the main  ESP32.

**Reply by 2xWhiskey** [reply]
The ELRS wifi network is being flakey with my desktop.  I don't think the hardware config took as when checking ure/hardware I get:

     "File Not Found

     URI: /hardware
     Method: GET
     Arguments: 0"

Reconnecting via wifi does not bring back the hardware config screen so I figure I need to reflash, but the same results, the standard homepage.  Is there a way to maybe completely wipe the ESP32 memory for a flash that will trigger the hardware config page again?  I tried the trashcan symbol next to the upload button, but it didn't help.

I have a tablet ready to for a better network connection.  I think my metal desk is interfering with the desktop PC's ability to use it.

**Reply by pkendall64** [reply]
web page is on /hardware.html and the json can be gotten from /hardware.json
You need the extension at the end.

**Reply by pkendall64** [reply]
The wifi is generally fairly weak on these things as the antenna is a pretty poor design.

**Reply by 2xWhiskey** [reply]
Wow, I feel like we must have gone through absolutely everything possible.  I just verified that yes the hardware config settings took, but the module is still refusing to communicate with the radio.

**Reply by pkendall64** [reply]
One thing you could try while debugging is to change 
```
    "serial_rx": 13,
    "serial_tx": 13,
```
and add in 
```
    "use_backpack": false,
    "debug_backpack_baud": 460800,
    "debug_backpack_rx": 3,
    "debug_backpack_tx": 1,
```
Then you can use a serial monitor to view what it prints when it boots. Also, reflash with -DDEBUG_LOG in super_defines.txt
You should see something like the following printed out
```
Initialised devices
Setting ISM2G4 Mode
Number of FHSS frequencies = 80
Sync channel = 41
41 37 20 45 28 64 16 66 71 25
39 57 14 6 55 44 35 9 31 76
69 47 19 29 4 1 40 24 46 12
48 3 33 17 62 43 26 0 36 67
22 56 11 78 58 27 30 68 23 2
13 50 63 77 70 5 18 49 79 34
74 7 61 38 8 53 65 52 59 15
42 75 54 51 10 72 21 73 60 32
41 56 39 70 46 59 32 9 57 53
14 44 66 11 35 26 58 29 30 17
78 25 40 8 5 36 69 63 43 13
45 28 74 0 75 42 31 27 6 61
47 64 7 60 3 62 16 1 23 10
4 73 20 21 72 77 19 33 22 67
38 68 65 55 48 15 18 24 76 50
2 49 51 79 54 12 52 34 71 37
41 18 48 12 5 27 71 34 9 22
36 13 38 68 59 54 43 4 32 55
16 6 26 47 75 11 0 17 57 15
25 23 29 58 21 64 70 63 33 46
28 66 39 3 45 14 35 42 61 78
72 73 40 74 2 7 8 60 79 20
56 1 24 67 69 51 37 65 31 52
19 49 62 10 77 53 30 76 50 44

ExpressLRS TX Module Booted...
Config version 7
Hal Init
Use TX pin: 33
Use RX pin: 32
SX1280 Reset
SX1280 Ready!
SX1280 Begin
Read Vers sx1280 #1: 43447
SetPower: 0
SetPower: 5
```

**Reply by 2xWhiskey** [reply]
Thanks, I will start on that tomorrow.  It's time for a recharge.

**Reply by 2xWhiskey** [reply]
I'm just seeing a bunch of garbled stuff in the terminal.  Shouldn't -DDEBUG_LOG be used in the user_defines.txt?
![Junk](https://user-images.githubusercontent.com/113867350/192383625-4faa4141-157f-4215-bff0-7c0d1dff6f39.jpg)

**Reply by pkendall64** [reply]
`super_defines.txt` is an override for `user_defines.txt` so we don't commit changes to `user_defines.txt`.
You might need to check the monitor_baud in unified.ini and change it to 460800.

**Reply by 2xWhiskey** [reply]
Ok, this time I get one readable message at the end: Setting ISMG4 Mode
![Junk](https://user-images.githubusercontent.com/113867350/192406652-0298b908-ca09-41fa-90c6-aa39f0138e2c.jpg)

**Reply by pkendall64** [reply]
If thats the only message then it's not compiled in debug mode with the -DDEBUG_LOG flag.

**Reply by 2xWhiskey** [reply]
Ok, I was unable to find super_defines.txt and had erroneously placed -DDEBUG_LOG into .gitignore.  I've been using the search tool but it doesn't seem to be finding the file.  Where is it supposed to be?

**Reply by pkendall64** [reply]
It should be a new file next to user_defines.txt, as I said it's used as an override so we don't need to make changes to user_defines.txt

**Reply by 2xWhiskey** [reply]
More issues:
![TooMany](https://user-images.githubusercontent.com/113867350/192421406-73f43c0c-b0df-4e7d-817e-a0221f171a07.jpg)
So then I changed the monitor_speed in unified.ini to 400000 and I get:
![TooMany2](https://user-images.githubusercontent.com/113867350/192421546-56211e26-c5a7-4c9f-91b5-73454e8e9c61.jpg)
Enabling the verbose define and going back to 420000 is also not helping.
Does UART inverted possibly have a relation?  Since it is selected by default and we didn't use a specific DupleTX .json?

**Reply by pkendall64** [reply]
The first image is good news! It means it's detected the radio chip properly and is talking to it.
The too many bad packets etc is because it's not communicating with the handset. Did you wire the handset to some different pins? and configure those as serial_tx and serial_tx? 
Can you post the json file for me.

**Reply by 2xWhiskey** [reply]
Hmm, I may not be using the correct handset pins, as I've been using what another guy provided to me for a pinout.  The guy who made YarrTX, aka brancell power user in Discord.  Hmm my handset has been powered off every time we have been doing this FYI.  I probably should disconnect my FTDI 3.3V before I power both up together.  I can't upload .json, so hardware.json:

{
    "customised":"true",
    "serial_rx":13,
    "serial_tx":13,
    "radio_busy":21,
    "radio_dio1":4,
    "radio_miso":19,
    "radio_mosi":23,
    "radio_nss":5,
    "radio_rst":14,
    "radio_sck":18,
    "power_rxen":27,
    "power_txen":26,
    "power_min":0,
    "power_high":4,
    "power_max":4,
    "power_default":2,
    "power_control":0,
    "power_values":[-15,-11,-8,-5,-1],
    "screen_type":0,
    "use_backpack": false,
    "debug_backpack_baud": 460800,
    "debug_backpack_rx": 3,
    "debug_backpack_tx": 1,
}

I'm going to call it a night, thanks as usual for your time!

**Reply by 2xWhiskey** [reply]
So with the serial monitor going and the FTDI connected to the module, I powered
 on the module via the handset in the specific model setup page. Here is what I got:
![Untitled](https://user-images.githubusercontent.com/113867350/192606306-c559b2f4-32e3-422d-a8dd-8dc52c05d6cb.jpg)
The garbles began when the module was powered up, power off just resulted in the same too many UART packet spam.

**Reply by 2xWhiskey** [reply]
With great help from **pkendall64**, we've finally gotten my module to work!  I plan on sharing my project in [ExpressLRS/ExpressLRS-Hardware](https://github.com/ExpressLRS
/
[ExpressLRS-Hardware](https://github.com/ExpressLRS/ExpressLRS-Hardware)) after a final PCB revision, and of course some personal testing with this working prototype board.

Thank you everyone for your patience and help!

**Comment by pkendall64** [comment]
I will move this to a discussion as it's not really an issue as such.

**Reply by 2xWhiskey** [reply]
Thank you for the very informative response!
